### ISSUE TITLE
load_weights fails for custom models

### ISSUE THREAD

**Issue by suth1807** [tensorflow/tensorflow#36362]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 18.04 LTS
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): Tensorflow 2.0
- Python version: 3.7.6
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: CPU 
- GPU model and memory:  CPU with 256GB RAM

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with: 1. TF 1.0: `python -c "import
tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"` 2. TF 2.0: `python -c
"import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
I created a simple custom model. Training runs fine and saving to hdf5 is also fine. When leading it back. I get the error
You are trying to load a weight file containing 3 layers into a model with 0 layers

**Describe the expected behavior**
The weights need to be loaded correctly

**Code to reproduce the issue**

import tensorflow as tf
from random import seed
from random import random

seed(1)
class SqDataset:
    def __init__(self):
        pass
    def generate(self):
        for i in range(1000):
            value = random()
            value = value * 2.0 - 1.0
            yield ({'x':value}, value*value)

dat = SqDataset()

datGen = tf.data.Dataset.from_generator(dat.generate, ({'x':tf.float32}, tf.float32), ({'x':tf.TensorShape([])}, tf.TensorShape([])))
datGen = datGen.batch(10)

class SquareWave(tf.keras.Model):
    def __init__(self):
        super(SquareWave, self).__init__()
        self.dense1 = tf.keras.layers.Dense(5, activation=tf.nn.tanh, name='sqdense1', input_shape=(1,))
        self.dense2 = tf.keras.layers.Dense(5, activation=tf.nn.tanh, name='sqdense2')
        self.dense3 = tf.keras.layers.Dense(1, activation=tf.nn.tanh, name='sqdense3')
        

    def call(self, inputs, training=False):
        ips = inputs['x']
        o = self.dense1(ips)
        o = self.dense2(o)
        o = self.dense3(o)
        return o

modelName = 'weights.h5'
if modelName != None:
    sqMode = SquareWave()
    sqMode.compile(optimizer=tf.keras.optimizers.RMSprop(), 
                   loss=tf.keras.losses.MeanSquaredError())
    sqMode.load_weights(modelName) #fails always
    
else:
    sqMode = SquareWave()
    sqMode.compile(optimizer=tf.keras.optimizers.RMSprop(), 
                   loss=tf.keras.losses.MeanSquaredError())
sqMode.fit(datGen, workers=0, epochs=100, use_multiprocessing=False)       

sqMode.save_weights('weights.h5', save_format='h5')

**Other info / logs**
The set of trainable_variables and trainable_weights is empty during the loading of model because the build methods are not called on the layers. This results in an error

**Comment by suth1807** [comment]
Just an update: SavedModel format works

**Comment by rmothukuru** [comment]
@suth1807,
Can you please share the Code with proper indentation and using Code Formatting Option,  **<>** shown above, or please execute the Code in the Google Colab and Save it as Github Gist (File -> Save a Copy as Github Gist), and please share that Link with us. Thanks!

**Comment by suth1807** [comment]
`import tensorflow as tf
from random import seed
from random import random

seed(1)
class SqDataset:
    def __init__(self):
        pass
    def generate(self):
        for i in range(1000):
            value = random()
            value = value * 2.0 - 1.0
            yield ({'x':value}, value*value)

dat = SqDataset()

datGen = tf.data.Dataset.from_generator(dat.generate, ({'x':tf.float32}, tf.float32), ({'x':tf.TensorShape([])}, tf.TensorShape([])))
datGen = datGen.batch(10)

class SquareWave(tf.keras.Model):
    def __init__(self):
        super(SquareWave, self).__init__()
        self.dense1 = tf.keras.layers.Dense(5, activation=tf.nn.tanh, name='sqdense1', input_shape=(1,))
        self.dense2 = tf.keras.layers.Dense(5, activation=tf.nn.tanh, name='sqdense2')
        self.dense3 = tf.keras.layers.Dense(1, activation=tf.nn.tanh, name='sqdense3')
        

    def call(self, inputs, training=False):
        ips = inputs['x']
        o = self.dense1(ips)
        o = self.dense2(o)
        o = self.dense3(o)
        return o



modelName = None#'weights.h5'
if modelName != None:
    sqMode = SquareWave()
    sqMode.compile(optimizer=tf.keras.optimizers.RMSprop(), 
                   loss=tf.keras.losses.MeanSquaredError())
    sqMode.load_weights(modelName) #fails always
    
else:
    sqMode = SquareWave()
    sqMode.compile(optimizer=tf.keras.optimizers.RMSprop(), 
                   loss=tf.keras.losses.MeanSquaredError())
sqMode.fit(datGen, workers=0, epochs=100, use_multiprocessing=False)       

sqMode.save_weights('weights.h5', save_format='h5')
pass`

**Comment by suth1807** [comment]
`import tensorflow as tf
from random import seed
from random import random

seed(1)
class SqDataset:
    def __init__(self):
        pass
    def generate(self):
        for i in range(1000):
            value = random()
            value = value * 2.0 - 1.0
            yield ({'x':value}, value*value)

dat = SqDataset()

datGen = tf.data.Dataset.from_generator(dat.generate, ({'x':tf.float32}, tf.float32), ({'x':tf.TensorShape([])}, tf.TensorShape([])))
datGen = datGen.batch(10)

class SquareWave(tf.keras.Model):
    def __init__(self):
        super(SquareWave, self).__init__()
        self.dense1 = tf.keras.layers.Dense(5, activation=tf.nn.tanh, name='sqdense1', input_shape=(1,))
        self.dense2 = tf.keras.layers.Dense(5, activation=tf.nn.tanh, name='sqdense2')
        self.dense3 = tf.keras.layers.Dense(1, activation=tf.nn.tanh, name='sqdense3')
        

    def call(self, inputs, training=False):
        ips = inputs['x']
        o = self.dense1(ips)
        o = self.dense2(o)
        o = self.dense3(o)
        return o



modelName = None#'weights.h5'
if modelName != None:
    sqMode = SquareWave()
    sqMode.compile(optimizer=tf.keras.optimizers.RMSprop(), 
                   loss=tf.keras.losses.MeanSquaredError())
    sqMode.load_weights(modelName) #fails always
    
else:
    sqMode = SquareWave()
    sqMode.compile(optimizer=tf.keras.optimizers.RMSprop(), 
                   loss=tf.keras.losses.MeanSquaredError())
sqMode.fit(datGen, workers=0, epochs=100, use_multiprocessing=False)       

sqMode.save_weights('weights.h5', save_format='h5')
pass`

**Comment by suth1807** [comment]
The link is below
https://gist.github.com/suth1807/9f2e0ab73046f7b85aa86959458bc4cd#file-untitled0-ipynb

**Comment by suth1807** [comment]
``import tensorflow as tf
from random import seed
from random import random

seed(1)
class SqDataset:
    def __init__(self):
        pass
    def generate(self):
        for i in range(1000):
            value = random()
            value = value * 2.0 - 1.0
            yield ({'x':value}, value*value)

dat = SqDataset()

datGen = tf.data.Dataset.from_generator(dat.generate, ({'x':tf.float32}, tf.float32), ({'x':tf.TensorShape([])}, tf.TensorShape([])))
datGen = datGen.batch(10)

class SquareWave(tf.keras.Model):
    def __init__(self):
        super(SquareWave, self).__init__()
        self.dense1 = tf.keras.layers.Dense(5, activation=tf.nn.tanh, name='sqdense1', input_shape=(1,))
        self.dense2 = tf.keras.layers.Dense(5, activation=tf.nn.tanh, name='sqdense2')
        self.dense3 = tf.keras.layers.Dense(1, activation=tf.nn.tanh, name='sqdense3')
        

    def call(self, inputs, training=False):
        ips = inputs['x']
        o = self.dense1(ips)
        o = self.dense2(o)
        o = self.dense3(o)
        return o



modelName = None#'weights.h5'
if modelName != None:
    sqMode = SquareWave()
    sqMode.compile(optimizer=tf.keras.optimizers.RMSprop(), 
                   loss=tf.keras.losses.MeanSquaredError())
    sqMode.load_weights(modelName) #fails always
    
else:
    sqMode = SquareWave()
    sqMode.compile(optimizer=tf.keras.optimizers.RMSprop(), 
                   loss=tf.keras.losses.MeanSquaredError())
sqMode.fit(datGen, workers=0, epochs=100, use_multiprocessing=False)       

sqMode.save_weights('weights.h5', save_format='h5')
pass``

**Comment by suth1807** [comment]
The code can be executed with modelName set to None to enable saving of the weights

**Comment by rmothukuru** [comment]
Could reproduce the error with Tensorflow Version 2.0. Here is the [Gist](https://colab.sandbox.google.com/gist/rmothukuru/1b3447627937fd6759f7a1dd235b5fde/36362.ipynb). Thanks!

**Comment by jvishnuvardhan** [comment]
@suth1807 There is a difference between subclass model and other keras model types(Sequential and Functional). Due to those differences listed [here](https://www.tensorflow.org/guide/keras/save_and_serialize#saving_subclassed_models), saving and loading of subclass models is different from other Keras models.

To restore a subclass model,  (1) you will need access to the code that created the model object and create the model again, (2) compile the model in order to restore the optimizer state and any stateful metrics, and (3) call it on some data before calling load_weights. 

You have followed steps (1) and (2) correctly and missed calling the compiled model on some data before loading weights using `load_weights`.

When I updated your code with the missing part of step(3), then everything worked as expected.

I am closing as it was resolved. Please feel free to reopen if the issue persists again. Thanks!

[Here](https://colab.sandbox.google.com/gist/jvishnuvardhan/2acfb877e7ba068f00c0294649524068/36362.ipynb) is the gist for y/our reference. Thanks again for the creating issue with good details.

**Comment by suth1807** [comment]
Thanks for the explanation! I had used this approach of making a dummy call
but thought of it as a work-around. I could also get the code to work
without making a dummy call by passing the correct data types
as shown below. In my opinion, this is a cleaner way to do it if the shapes
and data types are known. Do you agree?

        self.dense1 = tf.keras.layers.Dense(5, activation=tf.nn.tanh,
name='sqdense1', input_shape=(1,), dtype='float32')
        self.dense2 = tf.keras.layers.Dense(5, activation=tf.nn.tanh,
name='sqdense2',  dtype='float32')
        self.dense3 = tf.keras.layers.Dense(1, activation=tf.nn.tanh,
name='sqdense3',  dtype='float32')

**Comment by suth1807** [comment]
Thanks for the explanation! I had used this approach of making a dummy call but thought of it as a work-around. The input_shapes for the first layer is passed as argument so all the required information is there, isn't it?

**Comment by jvishnuvardhan** [comment]
@suth1807 `input_shape` in the first layer is required to infer shapes in Sequential and Functional keras model APIs. For subclass models, we need to call it on some data to initialize weights and then load weights. Hope this helps. Thanks!
